aten/src/ATen/native/TensorIterator.cpp

```diff
@@ -475,6 +475,7 @@ std::unique_ptr<TensorIterator> TensorIterator::reduce_op(Tensor& out, const Ten
   auto builder = TensorIterator::Builder();
   builder.add_output(out);
   builder.add_input(a);
+  builder.dont_compute_common_dtype();
```

Comment:
I'm worried that this will fail on CPU and fail on mixed-types that aren't fp16/fp32.

Perhaps this should only be called if out is fp32 and a is fp16 and both are CUDA.

Reply:
I don't think this is necessary for unary ops (hence reduction).

As I understand, reduce_op takes 1 input and 1 output, while it requires the output type to be the dtype for the operation. compute_common_dtype_ marks the promoted type operand types (`operands_[i].type`) which the op desired.
But since we are always converting the input to the desired output type (except for fp16/fp32 on gpu) https://github.com/pytorch/pytorch/pull/14580/files#diff-847680695fb0535da86db8b98a13b5a3R107, this check seems to be unnecessary.

But I can definitely add a check here if the lack of it makes us nervous.

Reply:
OK, I think you're right. But make sure there are unit tests for these cases.

Reply:
Test cases have been added.
Also put compute_common_dtype logic back in and added another flag for type promotion from cuda.HalfTensor to cuda.FloatTensor